Add support for hiding/unhiding objects

dodumosu opened the pull request:
This pull request addresses nditech/apollo-issues#171 by allowing a user to hide or show location sets, participant sets, events and forms.

dodumosu committed:
feat: add `is_hidden` field to models

dodumosu committed:
feat: add data migration

dodumosu committed:
feat: update list view and template

this commit adds the ability to toggle hidden status
to the admin view

dodumosu committed:
feat: remove hidden forms and events from UI

dodumosu committed:
feat: form list hiding (first version)

dodumosu committed:
fix: ensure toggles work, backend and front end

dodumosu committed:
feat: make RR and QA dashboards inaccessible
for hidden forms

dodumosu committed:
chore: change logic

dodumosu committed:
chore: update fixtures

dodumosu committed:
fix: update fixtures

dodumosu committed:
fix: update fixtures

dodumosu committed:
chore: remove unused import

dodumosu committed:
feat: reorder migrations and fix rendering

this commit reorders the migrations.
it also fixes issues with event display

dodumosu committed:
feat: don't show hidden events
for checklist/survey generation

takinbo commented on migrations/versions/14d0310ab80f_add_is_hidden_field.py:
```diff
@@ -0,0 +1,53 @@
+    )
+    op.add_column(
+        "participant_set", sa.Column("is_hidden", sa.Boolean(), nullable=True)
+    )
```
There's no point in making these columns nullable and they should have a default value of `False`

dodumosu replied:
all set to `default=False` and `nullable=False` on the respective model classes

takinbo commented on apollo/submissions/views_submissions.py:
```diff
@@ -1392,6 +1392,7 @@ def submission_version(submission_id, version_id):
 def quality_assurance_dashboard(form_id):
     form = services.forms.get_or_404(
         models.Form.id == form_id,
+        models.Form.is_hidden == False, # noqa
         models.Form.form_type.in_(['CHECKLIST', 'SURVEY']))
     breadcrumbs = [_('Quality Assurance Dashboard'), form.name]
     filter_class = filters.generate_quality_assurance_filter(g.event, form)
```
We should probably also have the same for the submissions list (Checklist, Surveys and Critical Incidents)

takinbo commented on apollo/formsframework/fixtures/38fff911ea3.sql:
```diff
@@ -8,15 +8,15 @@ INSERT INTO public.resource VALUES (6, 'form', 1, 'a0efcdf3-8f09-4d15-8303-6c720
+INSERT INTO public.form VALUES (3, 'IN-2-1', false, 'ZA', 'INCIDENT', true, true, NULL, '5e50b822cbf94fedb76619f1808716b2', 6, NULL, NULL, NULL, NULL, false, NULL, NULL, NULL);
+INSERT INTO public.form VALUES (4, 'CH-3-1', false, 'WA', 'CHECKLIST', true, true, NULL, 'aac237a0844d414e8635ed054ab27333', 7, NULL, NULL, NULL, NULL, false, NULL, NULL, NULL);
+INSERT INTO public.form VALUES (5, 'CH-3-2', false, 'WB', 'CHECKLIST', true, true, NULL, '64f973d56a55469da34ea3ec6b7ae2c8', 8, NULL, NULL, NULL, NULL, false, NULL, NULL, NULL);
+INSERT INTO public.form VALUES (6, 'IN-3-1', false, 'WC', 'INCIDENT', true, true, NULL, '34d79b36acf74dd99bbf8b68631f587c', 9, NULL, NULL, NULL, NULL, false, NULL, NULL, NULL);
```
This presumes to know the order in which the columns were created. To remove the ambiguity, the column names should be specified before the values.

takinbo commented on apollo/formsframework/views_forms.py:
```diff
@@ -215,25 +216,59 @@ def edit_form(view, form_id):
+    show_hidden = bool(query_params.get(show_hidden_param))
+
+    if show_hidden:
+        show_toggle_link_label = _('Hide Hidden')
```
What does "Hide Hidden" mean? The language should probably be clearer than this.

dodumosu replied:
changed to `Hide Archived`

takinbo commented on apollo/formsframework/views_forms.py:
```diff
@@ -215,25 +216,59 @@ def edit_form(view, form_id):
+        add_show_url_param = False
+    else:
+        show_toggle_link_label = _(
+            'Show All (%(count)d Hidden)', count=hidden_form_count)
```
Can you explain the rational for having the number of hidden objects displayed here?

dodumosu replied:
It was to give the user the number of items that weren't being displayed, but it's been removed

takinbo commented on apollo/frontend/views_admin.py:
```diff
@@ -103,6 +105,72 @@ def pre_validate(self, form):
+            {'is_hidden': True}, synchronize_session=False)
+        db.session.commit()
+
+    @action('unhide', _('Always Show'), _('Are you sure you want to always show the selected items?'))
```
Instead of using `Hide` and `Always Show`, could we instead use `Archive` and `Unarchive`?

dodumosu replied:
This has been changed

takinbo commented on apollo/templates/admin/form_list.html:
```diff
@@ -17,6 +17,9 @@
   </nav>
   <div class="d-flex justify-content-between align-items-start flex-row mt-1 mb-1">
     <div class="btn-toolbar d-none d-md-flex" role="toolbar">
+      <button class="btn btn-secondary mr-2 invisible" id="hide_button">{{ _('Hide Selected') }}</button>
+      <button class="btn btn-secondary mr-2 invisible" id="show_button">{{ _('Show Selected') }}</button>
```
`Archive` and `Unarchive`?

dodumosu replied:
Changed to `Archive` and `Unarchive`, respectively

dodumosu committed:
fix: specify columns

dodumosu committed:
feat: explicitly make columns not nullable

dodumosu committed:
chore: update migration so columns are not nullable

dodumosu committed:
chore: change labels

dodumosu committed:
chore: don't show submissions for archived forms

takinbo commented:
<img width="485" alt="Screenshot 2024-05-01 at 14 10 20" src="https://github.com/nditech/apollo/assets/41004/b2364bae-6c12-4465-a438-5d3ef5286e52">

@dodumosu can you update the migration so it's applied after `7cbcd7443117`? Given there are no conflicts of course.

dodumosu committed:
chore: reorder migrations

dodumosu commented:
done

> <img alt="Screenshot 2024-05-01 at 14 10 20" width="485" src="https://private-user-images.githubusercontent.com/41004/327129153-b2364bae-6c12-4465-a438-5d3ef5286e52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1ODIzODcsIm5iZiI6MTcxNDU4MjA4NywicGF0aCI6Ii80MTAwNC8zMjcxMjkxNTMtYjIzNjRiYWUtNmMxMi00NDY1LWE0MzgtNWQzZWY1Mjg2ZTUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDE2NDgwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjYmVmZDZmZjE3YTYyMGM1Y2NlY2U3OWNiMmQ5ODU2NWY2MjE0OGYzMWIyZTQ5N2I1NzQ1NzUxZmI0ZGU4NTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xQHB57tJs7_GYEHOndRXc3O1K41iwqrPRFCeyogAexs">
> @dodumosu can you update the migration so it's applied after `7cbcd7443117`? Given there are no conflicts of course.

done

dodumosu committed:
fix: set the server_default option on archivable models

takinbo commented:
Trying to edit any of the events (whether archived or not), yields this error:

<img width="911" alt="Screenshot 2024-05-13 at 17 15 01" src="https://github.com/nditech/apollo/assets/41004/6ac222ef-e91f-4490-a1ef-429fcc327794">

dodumosu committed:
fix: use correct filter method

dodumosu committed:
fix: workaround for flask-admin

this commit fixes an issue noticed by @takinbo.
it's caused by flask-admin using the `resource_id` field as the primary key for events, not the `id` field

dodumosu committed:
chore: minor refactor

takinbo commented:
There are still issues with the PR:

1. Hiding/Unhiding events does not work - they are still being displayed and I have no way of knowing if the event was hidden
2. Hiding a participant set does not remove it from available options when creating an event
3. Hiding a location set does not remove it from available options when creating a participant set
4. There's no way to archive forms

<img width="1225" alt="Screenshot 2024-05-14 at 04 50 09" src="https://github.com/nditech/apollo/assets/41004/e11a6701-7635-4c01-ae74-9f946cc3778d">

dodumosu committed:
fix: override parent method in Event admin class

dodumosu committed:
chore: add hidden icon, remove default checked status on archived forms

dodumosu committed:
feat: add custom form converter

dodumosu committed:
feat: remove hidden location sets when creating/editing participant sets

dodumosu committed:
chore: refactor

takinbo commented:
Archive/Unarchive buttons do not show up when the "All Forms" checkbox is selected.

![Screenshot 2024-05-15 at 20 43 46](https://github.com/nditech/apollo/assets/41004/cf57c6fc-9a82-4fbd-8d09-d6b7337b9661)

dodumosu commented:
They remain hidden until one or more check boxes are selected. Should I
change that behaviour?

On Wed, 15 May 2024, 20:44 Tim Akinbo, ***@***.***> wrote:

> Archive/Unarchive buttons do not show up when the "All Forms" checkbox is
> selected.
>
> Screenshot.2024-05-15.at.20.43.46.png (view on web)
> <https://github.com/nditech/apollo/assets/41004/cf57c6fc-9a82-4fbd-8d09-d6b7337b9661>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nditech/apollo/pull/884#issuecomment-2113333524>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKX5FXAVAJJGXPIZNFUE63ZCO3LDAVCNFSM5ULQG2NKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMJRGMZTGMZVGI2A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dodumosu committed:
fix: archival buttons should respond to toggle all checkbox

dodumosu committed:
feat: various changes

- hidden events cannot be switched to by admins anymore
- the default event processing skips hidden events
- role based access to resources excludes hidden ones (events and forms)

takinbo reviewed (approved):
tACK
